### ISSUE TITLE
Spurious (?) in-place add in batch norm trace

### ISSUE THREAD

**Issue by asuhan** [pytorch/pytorch#14826]
## 🐛 Bug

Batch norm trace contains an in-place add which appears to be related to recent in-place JIT changes.

## To Reproduce

```
from __future__ import print_function

import torch
import torch.nn as nn


class BatchNormModel(nn.Module):
    def __init__(self):
        super(BatchNormModel, self).__init__()
        self.bn = nn.BatchNorm2d(3)

    def forward(self, x):
        return self.bn(x)


x = torch.rand(14, 3, 5, 7)
model = BatchNormModel()
traced_model = torch.jit.trace(model, (x,))
traced_forward_graph = traced_model._get_method('forward').graph
print(traced_forward_graph)
```

The graph:

```
graph(%input : Float(14, 3, 5, 7)
      %weight : Float(3)
      %bias : Float(3)
      %running_mean : Float(3)
      %running_var : Float(3)
      %5 : Long()) {
  %6 : Long() = prim::Constant[value={1}](), scope: BatchNormModel/BatchNorm2d[bn]
  %7 : int = prim::Constant[value=1](), scope: BatchNormModel/BatchNorm2d[bn]
  %8 : Long() = aten::add_(%5, %6, %7), scope: BatchNormModel/BatchNorm2d[bn]
  %25 : bool = prim::Constant[value=1](), scope: BatchNormModel/BatchNorm2d[bn]
  %26 : float = prim::Constant[value=0.1](), scope: BatchNormModel/BatchNorm2d[bn]
  %27 : float = prim::Constant[value=1e-05](), scope: BatchNormModel/BatchNorm2d[bn]
  %28 : bool = prim::Constant[value=1](), scope: BatchNormModel/BatchNorm2d[bn]
  %29 : Float(14, 3, 5, 7) = aten::batch_norm(%input, %weight, %bias, %running_mean, %running_var, %25, %26, %27, %28), scope: BatchNormModel/BatchNorm2d[bn]
  return (%29);
}
```

## Expected behavior

Until recently (at least https://github.com/pytorch/pytorch/commit/1ca0ec7299b9352123a07525320a04e39d657781), the graph didn't contain the in-place `add_`:

```
graph(%input : Float(14, 3, 5, 7)
      %weight : Float(3)
      %bias : Float(3)
      %running_mean : Float(3)
      %running_var : Float(3)
      %5 : Long()) {
  %25 : bool = prim::Constant[value=1](), scope: BatchNormModel/BatchNorm2d[bn]
  %26 : float = prim::Constant[value=0.1](), scope: BatchNormModel/BatchNorm2d[bn]
  %27 : float = prim::Constant[value=1e-05](), scope: BatchNormModel/BatchNorm2d[bn]
  %28 : bool = prim::Constant[value=1](), scope: BatchNormModel/BatchNorm2d[bn]
  %29 : Float(14, 3, 5, 7) = aten::batch_norm(%input, %weight, %bias, %running_mean, %running_var, %25, %26, %27, %28), scope: BatchNormModel/BatchNorm2d[bn]
  return (%29);
}
```

## Environment

```
Collecting environment information...
PyTorch version: 1.0.0a0+524574a
Is debug build: Yes
CUDA used to build PyTorch: Could not collect

OS: Debian GNU/Linux
GCC version: (Debian 7.3.0-5) 7.3.0
CMake version: version 3.12.2

Python version: 3.5
Is CUDA available: No
CUDA runtime version: 9.2.148
GPU models and configuration: GPU 0: Quadro K1200
Nvidia driver version: 390.87
cuDNN version: Probably one of the following:
~/.cache/bazel/_bazel_asuhan/5e4d2e2a4156818f08a9e775a9d54463/external/local_config_cuda/cuda/cuda/lib/libcudnn.so
~/.cache/bazel/_bazel_asuhan/866353aec023bd55092aeb2e8f0f5ab4/external/local_config_cuda/cuda/cuda/lib/libcudnn.so
~/.cache/bazel/_bazel_asuhan/9886378d008b5a2d554b47b92d4d4999/external/local_config_cuda/cuda/cuda/lib/libcudnn.so
~/.cache/bazel/_bazel_asuhan/a8fe574dc2f657de728b2287a6b57d53/external/local_config_cuda/cuda/cuda/lib/libcudnn.so
~/.cache/bazel/_bazel_asuhan/d4cd244bdad0e09293d72b1af889e54d/external/local_config_cuda/cuda/cuda/lib/libcudnn.so

Versions of relevant libraries:
[pip3] numpy (1.12.1)
[conda] torch                     1.0.0a0+524574a           <pip>
[conda] torchvision               0.2.1                     <pip>
```

**Comment by SsnL** [comment]
I'm half-certain that the new trace is correct because the `add_` is probably on `num_batches_tracked` and it should be there. @suo, can you confirm this?

Also, there was a problem with traced BN because the running_mean & running_var are updated in-place but do not have their version counters bumped (see https://github.com/pytorch/pytorch/issues/13402) but it can still be traced by JIT, which at the time doesn't support in-place ops. I'm not sure if it will cause incorrect behavior or if it has been fixed by the mutability support. It'd be great if you can shed some light on this too @suo. :)

**Comment by suo** [comment]
Yes, it is tracing [this line](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/batchnorm.py#L67). I'm pretty sure the previous version (incorrectly) lowered the operation to the equivalent of `new_guy = self.num_batches_tracked + 1`, and then the entire op was DCE'd out, since `new_guy` was never explicitly used in the graph. This was a bug, since it made the counter increment a no-op. 

@SsnL it may fix the issue you linked, but I don't know much about the code; I'll reach out tomorrow to confirm.

**Comment by asuhan** [comment]
Ok, we'll work around this in our integration then.
